### PR TITLE
[Relay] Check match expressions for completeness

### PR DESCRIPTION
**Pull request by slyubomirsky**
We never defined the semantics of an incomplete match in Relay, so I've written a fairly simple function to check whether a match is complete (there's probably a better algorithm, but this one also concretely finds all cases that are not handled) and added into the type checker.

~This did require me to change several functions that relied on having incomplete matches, such as `hd` and `tl`. I do not think we should permit partial functions, as that would require us to define semantics for incomplete matches and potentially, down the line, to complicate the language with other kinds of error handling that could be placed higher in the stack (e.g., it would be a problem if some execution settings for Relay had one kind of error for incomplete matches and others reported the error a different way).~

~I suppose, though, that it would be easy to change the match incompleteness to a warning rather than an error if we really do want to keep `hd`, `tl`, etc.~

Incomplete matches only generate a warning for now, though we may want to add flags to make this an error later, see @MarisaKirisame's comment below

### CODE REVIEW

**Comment by MarisaKirisame** [comment]
@slyubomirsky can you make it LOG(WARNING) if a pattern is not exhausted right now? I think we should touch the hd/nth/tl function later.
So I think there is 3 pr:
0: the exhaustiveness checker that warn stuff
1: a flag 'strict' in the match node that check for exhaustiveness, and possibly redundancy later
2: discuss what to do with the partial functions, as @icemelon9 is using them.
I think we can take a similar approach as http://hackage.haskell.org/package/safe: provide version that is easiest to use, provide one that return default, provide one that return Maybe, etc.

**Comment by slyubomirsky** [comment]
I agree, I'll make it a warning

**Comment by MarisaKirisame** [comment]
@slyubomirsky http://moscova.inria.fr/~maranget/papers/warn/index.html
this paper conver redundancy as well. It's exhaustiveness algorithm is different. I think we can keep ours as it is, unless it get too slow.

**Comment by tqchen** [comment]
@MarisaKirisame please https://docs.tvm.ai/contribute/code_review.html#approve-and-request-changes-explicitly

**Comment by MarisaKirisame** [comment]
one small nitpick, otherwise is good.

File: src/relay/pass/match_exhaustion.cc [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    if (!op->constructor.same_as(ctor_cand->constructor)) {
+      return MatchResult::kClash;
+    }
+
+    // now check that subpatterns match
+    CHECK(op->patterns.size() == ctor_cand->patterns.size());
+    for (size_t i = 0; i < op->patterns.size(); i++) {
+      MatchResult submatch = this->Check(op->patterns[i], ctor_cand->patterns[i]);
+      if (submatch != MatchResult::kMatch) {
+        return submatch;
```
[the remark this answers is not in the document]
Comment (slyubomirsky) [reply src/relay/pass/match_exhaustion.cc]: Good point, I'll do that.

**Comment by slyubomirsky** [comment]
Pinging @MarisaKirisame and @icemelon9 for final review

**Pull request merged**